### PR TITLE
[Merged by Bors] - feat: `#parse` -- a command to parse text and log outputs

### PR DESCRIPTION
**Pull request by adomani**
This is a spin-off of #16230.  It introduces a function written by Matt that allows to use `#guard_msgs` to check parsing errors.

The new command is called `#guard_exceptions` and mimics very closely `#guard_msgs`.  As a way of showing how to use it, the PR also adds tests for Stacks Project Tags.

---

- [x] depends on: #16230

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Util/CaptureException.lean [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  let ictx := mkInputContext input "<input>"
+  let s := s.run ictx { env, options := {} } (getTokenTable env) (mkParserState input)
+  if !s.allErrors.isEmpty then
+    .error (s.toErrorMsg ictx)
+  else if ictx.input.atEnd s.pos then
+    .ok s.stxStack.back
+  else
+    .error ((s.mkError "end of input").toErrorMsg ictx)
+
+/-- `#guard_exceptions in` is based on `#guard_msgs in`, except that it captures parsing exceptions.
```
Comment (kmill) [inline Mathlib/Util/CaptureException.lean:28]: I propose that we redesign this a bit. Let's have a `#parse` command that tests a ParserFn:
```
"#parse " ident " => " str
```
Then, in test files, you would write `#guard_msgs in #parse Mathlib.Stacks.stacksTagFn => "A05"`

Yes, it's a bit wordier, but (1) it's more composable and (2) I think this is more likely for someone to be able to guess roughly what this is for when they see it.

File: Mathlib/Util/CaptureException.lean [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import Lean.Elab.Command
+
+/-!
+# `#guard_exceptions` -- an analogue of `#guard_msgs` for parsers
+-/
+
+namespace Mathlib.GuardExceptions
+
+open Lean Parser Elab Command
+/-- `captureException env s input` uses the given `Environment` `env` to parse the `String` `input`
```
[the remark this answers is not in the document]
Comment (adomani) [reply Mathlib/Util/CaptureException.lean:16]: Added your sentence!  :smile:

File: Mathlib/Util/CaptureException.lean (hunk 1 above)
Comment (adomani) [reply Mathlib/Util/CaptureException.lean]: I like this suggestion very much, thanks!  I implemented what you are saying (I hope!) and it looks great!

**Review by adomani — COMMENTED** [review]
Kyle, thank you for your comments!

Let me know if you have further suggestions.

**Review by YaelDillies — APPROVED** [review]
Can you update the PR title?

maintainer delegate

File: Mathlib/Util/CaptureException.lean [hunk 3]
```diff
@@ -0,0 +1,55 @@
+/-
+Copyright (c) 2024 Matthew Robert Ballard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Matthew Robert Ballard, Damiano Testa
+-/
+
+import Lean.Elab.Command
+
+/-!
+# `#guard_exceptions` -- an analogue of `#guard_msgs` for parsers
```
Comment (adomani) [inline Mathlib/Util/CaptureException.lean:10]:
```suggestion
# `#parse` -- a command to parse text and log outputs
```

File: Mathlib/Util/ParseCommand.lean [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  let s := s.run ictx { env, options := {} } (getTokenTable env) (mkParserState input)
+  if !s.allErrors.isEmpty then
+    .error (s.toErrorMsg ictx)
+  else if ictx.input.atEnd s.pos then
+    .ok s.stxStack.back
+  else
+    .error ((s.mkError "end of input").toErrorMsg ictx)
+
+/-- `#parse parserFnId => str` allows to capture parsing exceptions.
+`parserFnId` is the identifier of a `parserFn` and `str` is the string that `parserFn` will parse.
```
Comment (kim-em) [inline Mathlib/Util/ParseCommand.lean:33]:
```suggestion
`parserFnId` is the identifier of a `ParserFn` and `str` is the string that `parserFn` will parse.
```

**Comment by kim-em** [comment]
bors d+

**Comment by adomani** [comment]
bors merge

**Pull request closed without merging**